### PR TITLE
datasource: Optimize datasource detection, fix bugs

### PR DESCRIPTION
**Pull request by holmanb**
```
datasource: Optimize datasource detection, fix bug
      
Commit d1ffbea556a06105 enabled skipping python datasource detection on
OpenStack when no other datasources (besides DataSourceNone) can be discovered.
This allowed one to override detection, which is a requirement for OpenStack
Ironic which does not advertise itself to cloud-init.
      
Since no further datasources can be detected at this stage in the code, this
pattern can be generalized to other datasources to facilitate troubleshooting
or providing a general workaround to runtime detection bugs.
      
Additionally, this pattern can be extended to kernel commandline datasource
definition. Since kernel commandline is highest priority of the
configurations, it makes sense to override python code datasource
detection as well.

Include an integration test on LXD for this behavior that configures kernel
commandline and reboots to verify that the specified datasource is forced. 
```

## Additional Context
Continues work in d1ffbea556a06105d1ade88b4143ad43f53692c4 and https://github.com/canonical/cloud-init/pull/2044.

## Test

only works with `PLATFORM = "lxd_vm"`
```
tox -e integration-tests -- tests/integration_tests/datasources/test_openstack.py
```

## Future Work

Docs will come in a separate PR. This will include documentation of the kernel commandline invocation required, as well as per-datasource keys.

### CODE REVIEW

File: tests/unittests/sources/test_cloudsigma.py [hunk 1]
```diff
@@ -56,6 +57,9 @@ def setUp(self):
         self.datasource.cepko = CepkoMock(SERVER_CONTEXT)
 
     def test_get_hostname(self):
+        # fails on python 3.6
```
Comment (TheRealFalcon) [inline tests/unittests/sources/test_cloudsigma.py:60]: Why?

File: tests/integration_tests/datasources/test_detect_openstack.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import pytest
+
+from tests.integration_tests.instances import IntegrationInstance
+
+
+@pytest.mark.lxd_vm
+@pytest.mark.lxd_container
```
Comment (TheRealFalcon) [inline tests/integration_tests/datasources/test_detect_openstack.py:7]: This can't actually run on lxd_container right?

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
I like this change and the PR overall looks really good. I have a few minor inline comments.

> Docs will come in a separate PR

Ticket number or I don't believe you :wink:

File: cloudinit/sources/__init__.py [hunk 3]
```diff
@@ -311,22 +312,43 @@ def ds_detect(self) -> bool:
         """Check if running on this datasource"""
         return True
 
+    @staticmethod
+    def parse_cmdline():
+        """Check if command line argument for this datasource was passed
+        Passing by command line overrides runtime datasource detection
+        """
+        cmdline = util.get_cmdline()
+        ds_parse_1 = re.search(r"ci\.ds=([a-zA-Z]+) ", cmdline)
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/sources/__init__.py:321]: > Can datasource be the last thing on the cli?

Yes

> r"ci\.ds=([a-zA-Z]+)( |$)"

+1 Nice suggestion.

> In my testing, it detects the space before the end of line, but I'm not sure if there's a better way to indicate that as taking precedence.

I don't follow what you're saying about precedence.

> Also does it make sense to capture more whitespace than a space?
Are you talking about different kinds of whitespace here? As in, if someone separates arguments with newlines or tabs or something weird like that?

File: cloudinit/sources/__init__.py [hunk 4]
```diff
@@ -311,22 +312,43 @@ def ds_detect(self) -> bool:
...
+        - commandline argument is used (ci.ds=OpenStack)
+
+        Note: get_cmdline() is required for the general case - when ds-identify
+        does not run, _something_ needs to detect the kernel command line
+        definition.
         """
-        return self.sys_cfg.get("datasource_list", []) in (
+        if self.dsname == self.parse_cmdline():
+            return (
+                "Machine is configured by the kernel commandline to run on "
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/sources/__init__.py:338]: > slightly more intuitive here

Heh, not sure what I was thinking here. Returning log messages as an implicit truthy value is not intuitive, agreed.

File: cloudinit/sources/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/sources/__init__.py:321]: > I don't follow what you're saying about precedence.

Yeah, nevermind...thinking about it again my question doesn't make sense.

> Are you talking about different kinds of whitespace here? As in, if someone separates arguments with newlines or tabs or something weird like that?

Correct

[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/sources/__init__.py:321]: > > Are you talking about different kinds of white space here? As in, if someone separates arguments with newlines or tabs or something weird like that?

> Correct

+1 I'll cook something up.

File: tests/unittests/sources/test_cloudsigma.py (hunk 1 above)
Comment (holmanb) [reply tests/unittests/sources/test_cloudsigma.py]: Nevermind, dropped these

**Comment by holmanb** [comment]
> > Docs will come in a separate PR
> 
> Ticket number or I don't believe you

[Sure](https://warthogs.atlassian.net/browse/SC-1457) :-)

I think I've addressed everything. If tests pass I think this is ready for another round of review.

**Review by TheRealFalcon — APPROVED** [review]
LGTM!

**Pull request merged**